### PR TITLE
Juno @task-5

### PR DESCRIPTION
**Pull request by PsychoPunkSage**
**NOTE**: Please don't merge it. It is an exercise PR

### CODE REVIEW

**Comment by PsychoPunkSage** [comment]
Hi @rianhughes 
I have created a fresh PR for task-5. please have a look here.
Thank You

**Comment by PsychoPunkSage** [comment]
I tried to implement `NodesFromRoot` Mock... but I'm facing little difficulty in doing so...
Can you please have a look at the code and tell me what I'm doing Wrong?? Or what changes I have to make??

File: rpc/handlers.go [hunk 1]
```diff
@@ -483,5 +489,35 @@ func (h *Handler) MethodsV0_6() ([]jsonrpc.Method, string) { //nolint: funlen
...
+			Params:  []jsonrpc.Parameter{{Name: "key"}},
+			Handler: h.JunoGetNodesFromRoot,
+		},
 	}, "/v0_6"
 }
+
+func (h *Handler) JunoGetNodesFromRoot(key felt.Felt) (string, *jsonrpc.Error) {
+	stateReader, _, error := h.bcReader.HeadState()
+	if error != nil {
+		return "", ErrBlockNotFound
```
Comment (rianhughes) [inline rpc/handlers.go:503]: You should return the actual error here

**Comment by rianhughes** [comment]
> I tried to implement `NodesFromRoot` Mock... but I'm facing little difficulty in doing so... Can you please have a look at the code and tell me what I'm doing Wrong?? Or what changes I have to make??

So the code inside the mocks folder is actually autogenerated. You should enter the core directory and run
go generate github.com/NethermindEth/juno/core
(VSCode also has a "go generate" button on top of the interface definition - I typically just use this). Note you might have to make juno first

File: rpc/handlers_test.go [hunk 2]
```diff
@@ -99,3 +99,22 @@ func TestThrottledVMError(t *testing.T) {
...
+	t.Cleanup(mockCtrl.Finish)
+	mockReader := mocks.NewMockReader(mockCtrl)
+	mockReader.EXPECT().Network().Return(&utils.Mainnet).AnyTimes()
+
+	log := utils.NewNopZapLogger()
+	mockState := mocks.NewMockStateHistoryReader(mockCtrl)
+
+	handler := rpc.New(mockReader, nil, nil, "", utils.Ptr(utils.Mainnet), log)
+
+	t.Run("Key DNE", func(t *testing.T) {
```
Comment (rianhughes) [inline rpc/handlers_test.go:114]: I would add another test where nodeStorage is not nil

File: rpc/handlers.go (hunk 1 above)
Comment (PsychoPunkSage) [reply rpc/handlers.go]: resolved

**Comment by PsychoPunkSage** [comment]
> > I tried to implement `NodesFromRoot` Mock... but I'm facing little difficulty in doing so... Can you please have a look at the code and tell me what I'm doing Wrong?? Or what changes I have to make??
> 
> So the code inside the mocks folder is actually autogenerated. You should enter the core directory and run go generate github.com/NethermindEth/juno/core (VSCode also has a "go generate" button on top of the interface definition - I typically just use this). Note you might have to make juno first

After entering `core` dir. I ran `go generate github.com/NethermindEth/juno/core` but i got this error::
```
state.go:28: running "mockgen": exec: "mockgen": executable file not found in $PATH
```
I think I need to add binary/.exe file or something to the PATH var.... I am not aware about it... can you tell how to fix this??

Plus, I didn't get the part `Make Juno first`... What do you mean by that?? I didn't get what I need to do cause I have already build Juno...

File: rpc/handlers_test.go (hunk 2 above)
Comment (PsychoPunkSage) [reply rpc/handlers_test.go:115]: Resolved

File: rpc/handlers.go [hunk 3]
```diff
@@ -483,5 +489,52 @@ func (h *Handler) MethodsV0_6() ([]jsonrpc.Method, string) { //nolint: funlen
...
 }
+
+func (h *Handler) JunoGetNodesFromRoot(key felt.Felt) (string, *jsonrpc.Error) {
+	stateReader, _, error := h.bcReader.HeadState()
+	if error != nil {
+		return "", &jsonrpc.Error{
+			Code:    11111,
+			Message: "Failed to retrieve head state",
+			Data:    error.Error(),
+		}
```
Comment (rianhughes) [inline rpc/handlers.go:507]: jsonrpc.Err(jsonrpc.InternalError, err.Error())

File: rpc/handlers_test.go [hunk 4]
```diff
@@ -99,3 +100,30 @@ func TestThrottledVMError(t *testing.T) {
...
+		mockReader.EXPECT().HeadState().Return(mockState, nopCloser, nil)
+		nodeStorage, err := handler.JunoGetNodesFromRoot(felt.Zero)
+		require.Nil(t, nodeStorage)
+		require.NotEqual(t, err, nil)
+	})
+
+	t.Run("Key Exist", func(t *testing.T) {
+		element := fp1.NewElement(1)
```
Comment (rianhughes) [inline rpc/handlers_test.go:123]: new(felt.Felt).SetUint64(1)

File: rpc/handlers_test.go (hunk 2 above)
Comment (rianhughes) [reply rpc/handlers_test.go:115]: It doesn't seem to run

File: rpc/handlers.go (hunk 3 above)
Comment (PsychoPunkSage) [reply rpc/handlers.go]: Resolved

File: rpc/handlers_test.go (hunk 4 above)
Comment (PsychoPunkSage) [reply rpc/handlers_test.go:123]: Hi @rianhughes ....
Is there any issue in the way I defined it?? 

```go
element := fp1.NewElement(1)
mockReader.EXPECT().HeadState().Return(mockState, nopCloser, nil)
nodeStorage, err := handler.JunoGetNodesFromRoot(*felt.NewFelt(&element))
```

Like its compiling so I thought why to replace it with `new(felt.Felt).SetUint64(1)`

Comment (PsychoPunkSage) [reply rpc/handlers_test.go]: Resolved

Comment (rianhughes) [reply rpc/handlers_test.go]: It's just neater (and safer?) to use the juno felt package, since it's being used everywhere else in the code.
